awx/main/validators.py

```diff
@@ -48,24 +48,32 @@ def validate_pem(data, min_keys=0, max_keys=None, min_certs=0, max_certs=None):
 
     # Build regular expressions for matching each object in the PEM file.
     pem_obj_re = re.compile(
-        r'^(-{4,}) *BEGIN ([A-Z ]+?) *\1[\r\n]+' +
```

Comment:
I want to focus on the first change here, going from `(-{4,})` to `-{5}\`. This doesn't seem related to the desired change here (to allow new lines). It's hypothetically possible that someone has some client code that uses 4 or greater than 5 `-` characters in their keys, and this would break that. Ultimately, we would like to adhere to the proper standards for PEM files. If you have some justification for this change, I think we can consider it, but without a reason for the change I would prefer to keep the old syntax of `-{4,}`.

I also realize that you are removing the capturing group because it was not being used. That's okay with me, I just need to be extra sure that we changed the numbering of the other groups in every place they are used.

Reply:
I also agree that it's probably more clear to not use the references, `\1` and `\2`, instead just putting in the intended pattern again. But it still makes it a little more difficult to get down to the main intended change here.

Reply:
I reintroduced `-{4,}` these would be used by the back reference group `\1` and `\2` would be the key type. 
I gave each match group a human readable name for clarity and double checked their usage. I've also run the unit tests a good few times.

Reply:
Oh I'm really sorry, I was wrong about the `/1` and `/2` references. The intention is that the number of dashes must be constant.

So

```
-----BEGIN PRIVATE KEY-----
foobar
-----END PRIVATE KEY----
```

should fail because the end has the wrong number of dashes

and the second backreference was protecting against cases like

```
-----BEGIN PRIVATE KEY-----
foobar
-----END PRIVATE CERTIFICATE-----
```

So that should fail too.

It looks like that's also important to make sense of crazy cases like

```
-----BEGIN PRIVATE KEY-----
foobar
-----END PRIVATE KEY----------BEGIN PRIVATE KEY-----
foobar
-----END PRIVATE KEY-----
```

Reply:
Ooohkay, sure, that does make sense. I was wondering. That test case also makes sense to me now. We can have named backreferences to.

Reply:
FYI, [RFC7468](https://tools.ietf.org/html/rfc7468) specifies 5 dashes with no extra space (e.g.`-----BEGIN FOO-----`), however others like [RFC4716](https://tools.ietf.org/html/rfc4716) use 4 dashes with a space (e.g. `---- BEGIN SSH2 PUBLIC KEY ----`), which is why there's a little flexibility in dashes/spaces accepted.